corehq/motech/openmrs/finders.py

```diff
@@ -0,0 +1,165 @@
+        #         }
+        #       },
+        #       // ...
+        #     }
+        #
+
+        for person_prop, value_source in case_config['person_properties'].items():
+            jsonpath = 'person.' + person_prop
+            self.property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+        for attr_uuid, value_source in case_config['person_attributes'].items():
+            jsonpath = 'person.attributes.value where `parent`.attributeType.uuid = "' + attr_uuid + '"'
+            self.property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+        for name_prop, value_source in case_config['person_preferred_name'].items():
+            jsonpath = 'person.preferredName.' + name_prop
+            self.property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+        for addr_prop, value_source in case_config['person_preferred_address'].items():
+            jsonpath = 'person.preferredAddress.' + addr_prop
+            self.property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+    def get_score(self, patient, case):
+        """
+        Return the sum of weighted properties to give an OpenMRS
+        patient a score of how well they match a CommCare case.
+        """
+        def weights():
+            for prop, weight in self.property_weights.items():
+                case_value = case.get_case_property(prop)
+                jsonpath_expr = parse(self.property_map[prop].jsonpath)
+                patient_value = jsonpath_expr.find(patient)
+                value_map = self.property_map[prop].value_map
+                is_equal = value_map.get(patient_value, patient_value) == case_value
+                yield weight if is_equal else 0
+
+        return sum(weights())
+
+    def find_patients(self, requests, case, case_config):
+        """
+        Matches cases to patients. Returns a list of patients, each
+        with a confidence score >= THRESHOLD
+        """
+        PatientScore = namedtuple('PatientScore', ['patient', 'score'])
+        self.set_property_map(case_config)
+
+        candidates = {}  # key on OpenMRS UUID to filter duplicates
+        for prop in self.searchable_properties:
+            value = case.get_case_property(prop)
+            response = requests.get('/ws/rest/v1/patient', {'q': value, 'v': 'full'})
+            for patient in response.json()['results']:
+                score = self.get_score(patient, case)
+                if score >= self.threshold:
+                    candidates[patient['uuid']] = PatientScore(patient, score)
+        if not candidates:
+            return []
+        patients_scores = sorted(candidates.values(), key=lambda cand: cand.score, reverse=True)
+        if len(patients_scores) == 1:
+            return [patients_scores[0].patient]
+        if float(patients_scores[0].score) / patients_scores[1].score > 1 + self.confidence_margin:
```

Comment:
Yeah but no, Mr Stickler. That numerator is a float, so that's obviously [true division](https://www.python.org/dev/peps/pep-0238/).

Do you have any strong feelings about "future division" vs casting, reviewers?

Reply:
I think it's just saying you should include

```
from __future__ import division
```
https://github.com/janjur/readable-pylint-messages/blob/master/README.md#old-division

as a way to ensure that this remains python3 compatible. Even though in this case what you wrote here I think would be python3-compatible, not casting and using the `__future__` import marks this file's division as python3-compatible so when we make the switch no one has to think about whether this particular line is or isn't compatible. Does that make sense? I think in general that's why we turned on stricter py3-compat linter errors.